2025-05-99-how-to-implement-zero-copy-communication/index.md

```diff
@@ -0,0 +1,164 @@
+### What Is iceoryx2
+
+**iceoryx2** is a decentralized, service-based inter-process communication
+(IPC) library designed for mission-critical, high-efficiency systems. It
+achieves low latency and high throughput using a communication paradigm known
+as zero-copy communication.
+
+In this article, we'll explore the core concept of zero-copy communication and
+walk through how iceoryx2 implements it using a practical example from its
+publish-subscribe API.
+
+### The Concept of Zero-Copy Communication
+
+The principle of zero-copy communication is similar to how shared pointers or
+references work in programming: instead of copying data, we share access to a
+single memory location.
+
+In the same spirit, zero-copy inter-process communication avoids unnecessary
+data copies between processes. Data is produced once in a memory region that
+all endpoints can access, and a reference or offset to that memory is shared.
+This allows all participants to read the data directly, just like sharing a
+pointer to a struct on the heap - but across process boundaries.
+
+In practice, this shared memory region could be:
+
+* POSIX shared memory
+* GPU memory
+* Memory-mapped hardware buffers
+
+The critical aspect: the sender writes the data directly into shared memory.
+Then, the offset or index to that data is transferred to every receiver.
+No copies are involved after the data is created.
+
+<table width=100% height=100% border=0><tr><td align=center>
+
+<img src="shared-memory-idea.png" alt="shared-memory-idea" style="max-width: 50%"/>
+
+</td></tr></table>
+
+### Hands-On: iceoryx2 in Action
+
+We’ll now walk through the
+[iceoryx2 publish-subscribe example](https://github.com/eclipse-iceoryx/iceoryx2/tree/main/examples/rust/publish_subscribe),
+where data is sent by a publisher to multiple subscribers using zero-copy.
+
+Sensor APIs often provide a pointer to the latest data:
+
+```rust
+let data_ptr = sensor_get_next_frame();
+```
+
+However, this is not zero-copy: the data resides in sensor-managed memory, and
+must be copied into shared memory before sending. For true zero-copy, we
+need the sensor to write directly into shared memory, provided by the
+zero-copy communication framework.
+
+* **1. Produce Data into Shared Memory**
+  First, we acquire an uninitialized sample from the publisher. This is a chunk
+  of memory from the publisher’s data segment, reserved for the user's payload.
+  Under the hood, an allocator manages this memory block:
+
+  ```rust
+  let mut uninitialized_sample =
+      publisher.loan_slice_uninit(MAX_SIZE_OF_SENSOR_DATA_FRAME)?;
+  ```
+
+  Next, the sensor writes directly into this memory:
+
+  ```rust
+  sensor_get_next_frame(uninitialized_sample.payload_mut());
+  ```
+
+  * **Deep Dive:**
+    When the publisher is created:
+    1. A shared memory object is created using `shm_open`.
+    2. It is resized via `ftruncate` to reserve enough space.
+    3. The segment is mapped into the process using `mmap`.<br><br>
```

Comment:
This feels a bit out of place since the creation of the publisher is never mentioned before and the publisher is already used. What do you think about moving this to line 56.

Also, to stand out visually a bit more, what do you think about using quotation instead of a bullet point, e.g.

> **Deep Dive:**
> When ...

But I'm not sure if it looks good in a larger document.